core/Objects/NPMClient.cs

```diff
@@ -287,39 +287,156 @@ public async Task PublishPackage(IFileInfo package)
+                    "_attachments": {
+                        "{{tgz}}": {
+                        "content_type": "application/octet-stream",
+                        "data": "{{dataBase64}}",
+                        "length": {{package.Length}}
+                        }
+                    }
+                }
+                """);
+            // patch version manifest
+            root["versions"][ctrlr.CmfPackage.Version]["_id"] = $"{name}@{ctrlr.CmfPackage.Version}";
+            root["versions"][ctrlr.CmfPackage.Version]["_cliVersion"] = ExecutionContext.CurrentVersion;
+            root["versions"][ctrlr.CmfPackage.Version]["_integrity"] = sha512Hash;
+            root["versions"][ctrlr.CmfPackage.Version]["dist"] = JObject.Parse($$"""
+                    {
+                    "integrity": "{{sha512Hash}}",
+                    "shasum": "{{sha1Hash}}",
+                    "tarball": "{{this.baseUrl.Replace("https://", "http://")}}/{{name}}/-/{{tgz}}"
+                    }
+                    """);
 
+
+
+            var payload = root.ToString(Formatting.None);
+            var content = new StringContent(payload, Encoding.UTF8, "application/json");
+
+            return content;
+        }
+
+        private HttpContent GetPublishBodyStream(
+            IFileInfo package,
+            CmfPackageController ctrlr,
+            string name,
+            string tgz,
+            string manifest
+        )
+        {
+            const string JSON_DATA_PLACEHOLDER = "$$$TGZ_BASE_64_TOKEN$$$";
+
+            JObject root = null;
+
+            string sha1Hash;
+            {
+                using var fileStream = package.OpenRead();
                 using SHA1 sha1 = SHA1.Create();
-                byte[] sha1HashBytes = sha1.ComputeHash(fileBytes);
-                var sha1Hash = BitConverter.ToString(sha1HashBytes).Replace("-", "").ToLower();
+                byte[] sha1HashBytes = sha1.ComputeHash(fileStream);
+                sha1Hash = BitConverter.ToString(sha1HashBytes).Replace("-", "").ToLower();
+            }
 
+            string sha512Hash;
+            {
+                using var fileStream = package.OpenRead();
                 using SHA512 sha512 = SHA512.Create();
-                byte[] sha512HashBytes = sha512.ComputeHash(fileBytes);
+                byte[] sha512HashBytes = sha512.ComputeHash(fileStream);
                 var sha512_64 = Convert.ToBase64String(sha512HashBytes);
                 // var sha512Hash = BitConverter.ToString(sha512HashBytes).Replace("-", "").ToLower();
-                var sha512Hash = $"sha512-{sha512_64}";
-                Log.Debug($"Package content with hash {sha512Hash} and checksum {sha1Hash}");
-                
-                root = JObject.Parse(
+                sha512Hash = $"sha512-{sha512_64}";
+            }
+
+            Log.Debug($"Package content with hash {sha512Hash} and checksum {sha1Hash}");
+
+            root = JObject.Parse(
```

Comment:
we should probably extract this manifest to a function, as it is exactly the same in both streaming and non-streaming mode, except for the placeholder

Reply:
Yes, makes sense. Will do :)